host_source.go

```diff
@@ -271,106 +365,279 @@ func checkSystemLocal(control *controlConn) (bool, error) {
+	if it == nil {
+		return nil, errors.New("Attempted to query 'system.local' on a closed control connection")
+	}
 
-		if err = iter.Close(); err != nil {
-			return nil, "", err
-		}
-	} else {
-		iter := r.session.control.withConn(func(c *Conn) *Iter {
-			localHost = c.host
-			return c.query(legacyLocalQuery)
-		})
+	// expect only 1 row
+	it.MapScan(row)
+	if err := it.Close(); err != nil {
+		return nil, err
+	}
 
-		if iter == nil {
-			return r.prevHosts, r.prevPartitioner, nil
-		}
+	// extract all available info about the host
+	err, host := r.hostInfoFromMap(row)
+	if err != nil {
+		return nil, err
+	}
+
+	return host, err
+}
 
-		iter.Scan(&localHost.dataCenter, &localHost.rack, &localHost.hostId, &localHost.tokens, &partitioner, &localHost.version)
+// Given an ip address and port, return a peer that matched the ip address
+func (r *ringDescriber) GetPeerHostInfo(ip net.IP, port int) (*HostInfo, error) {
+	row := make(map[string]interface{})
 
-		if err = iter.Close(); err != nil {
-			return nil, "", err
-		}
+	it := r.session.control.query("SELECT * FROM system.peers WHERE peer=?", ip)
+	if it == nil {
+		return nil, errors.New("Attempted to query 'system.peers' on a closed control connection")
+	}
+
+	// expect only 1 row
+	it.MapScan(row)
+	if err := it.Close(); err != nil {
+		return nil, err
 	}
 
-	localHost.port = r.session.cfg.Port
+	// extract all available info about the host
+	err, host := r.hostInfoFromMap(row)
+	if err != nil {
+		return nil, err
+	}
+
+	return host, err
+}
 
-	hosts = []*HostInfo{localHost}
+// Ask the control node for host info on all it's known peers
+func (r *ringDescriber) GetClusterPeerInfo() ([]*HostInfo, error) {
+	var hosts []*HostInfo
 
-	rows := r.session.control.query("SELECT rpc_address, data_center, rack, host_id, tokens, release_version FROM system.peers").Scanner()
-	if rows == nil {
-		return r.prevHosts, r.prevPartitioner, nil
+	// Ask the node for a list of it's peers
+	it := r.session.control.query("SELECT * FROM system.peers")
+	if it == nil {
+		return nil, errors.New("Attempted to query 'system.peers' on a closed connection")
 	}
 
-	for rows.Next() {
-		host := &HostInfo{port: r.session.cfg.Port}
-		err := rows.Scan(&host.peer, &host.dataCenter, &host.rack, &host.hostId, &host.tokens, &host.version)
+	for {
+		row := make(map[string]interface{})
+		if !it.MapScan(row) {
+			break
+		}
+		// extract all available info about the peer
+		err, host := r.hostInfoFromMap(row)
 		if err != nil {
-			Logger.Println(err)
-			continue
+			return nil, err
 		}
 
-		if r.matchFilter(host) {
-			hosts = append(hosts, host)
+		// If it's not a valid peer
+		if !r.IsValidPeer(host) {
+			Logger.Printf("Found invalid peer '%+v' "+
+				"Likely due to a gossip or snitch issue, this host will be ignored", host)
+			continue
 		}
+		hosts = append(hosts, host)
+	}
+	if it.err != nil {
+		return nil, errors.Wrap(it.err, "GetClusterPeerInfo()")
```

Comment:
As Dave Cheney lays out here https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully annotated errors are useful for end users, but since gocql does not provide annotated errors; I agree, this doesn't make much sense here.